mono/eglib/glib.h

```diff
@@ -1133,10 +1139,37 @@ gboolean   g_file_test (const gchar *filename, GFileTest test);
 
 gchar *g_mkdtemp (gchar *tmpl);
 
-
 /*
  * Low-level write-based printing functions
  */
+
+static inline int
+g_async_safe_fgets (char *str, int num, int handle, gboolean *newline)
+{
+	memset (str, '\0', sizeof(char) * num);
+	// Make sure we always have null-termination
+	int without_padding = num - 1;
+	int i=0;
+	char scratch [2];
+	while (i < without_padding && g_read (handle, scratch, sizeof(char))) {
```

Comment:
Reading one char at a time is very inefficient, but I see helps correctness elsewhere.

Reply:
Suggest to do it with larger reads, but I know it takes more work.

Reply:
It's also the case that making the size of the stack frame larger than a very small amount is dangerous here. The signal handler code runs in an alternate stack, which is much smaller. It's easy to overflow that stack and crash the crash reporter. It's simply not safe to allocate more scratch space, and reading one char at a time ensures that I never need to "put back" any characters.

Reply:
Reading one byte at a time is extremely slow. I know it is crash reporting. Hopefully only ever that. The caller’s buffer could be better used, along with a position into it.

Reply:
Reading one byte at a time is extremely slow. I know it is crash reporting. Hopefully only ever that. The caller’s buffer could be better used, along with a position into it.

Reply:
I can definitely read into the caller's array, but reading by more than one byte at a time would prevent the reading from stopping the minute it found any nonprintable characters, and the algorithm for skipping them would have to change. 

I could come up with some logic of overwriting the caller's buffer's terminators with '\0', but then printing that would have to involve "restarting" printing directly after the skipped character. It seems prone to risk of error. 

It should be noted that producing invalid JSON here means that the crash reporting can't be parsed, so this is mission critical. Also noted is that as slow as I'm sure it is, it doesn't seem to have any perceptible impact on the crash dumper. This code has some other major perf killers, such as global thread synchronization using signals.

Reply:
I have changed this to use the caller's memory, but it still reads one char at a time. Let's wait until we see it being a perf problem before we try a less-reliable reading policy.